### ISSUE TITLE
[FEA] desloc_bench: Ablation — 各优化组件贡献

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#585]
## Summary

消融实验：逐一关闭 DES-LOC 的核心优化组件（local SGD、slow momentum、grad compression、adaptive K），单独量化每个组件对吞吐和收敛的贡献，并评估组件组合效果。

## Design Reference

- **File**: `benchmarks/desloc_ablation.py` (新建)
- **Parent EPIC**: #299 | **Parent THEME**: #276

## Background

DES-LOC 的性能收益来自多个协同机制：① local SGD 减少同步频次；② slow momentum 平滑 tier 间参数漂移；③ Top-K 梯度压缩降低通信量；④ adaptive K 根据带宽动态调整同步步频。缺少任一组件都可能导致收敛退化或通信瓶颈，但各组件的独立贡献此前缺乏系统量化。

本消融 benchmark 采用"一次关闭一个"策略，在统一 baseline（2H100+3A6000，GPT-125M，1000步）上对比各配置。

## Ablation Configs

| Config | local_sgd_K | slow_momentum | compression | adaptive_K |
|--------|-------------|---------------|-------------|------------|
| Full DES-LOC (baseline) | 5 | 0.9 | 5% | ✓ |
| w/o local SGD (K=1) | 1 | 0.9 | 5% | ✗ |
| w/o slow momentum | 5 | 0.0 | 5% | ✓ |
| w/o compression | 5 | 0.9 | 100% | ✓ |
| w/o adaptive K | 5 | 0.9 | 5% | ✗ |
| Minimal (all off) | 1 | 0.0 | 100% | ✗ |

## Implementation

```python
# benchmarks/desloc_ablation.py
"""
Ablation study: contribution of each DES-LOC component.

Components:
  - local_sgd   : K > 1 (reduce inter-tier sync frequency)
  - slow_mom    : slow momentum for inter-tier averaging
  - compression : Top-K gradient sparsification
  - adaptive_k  : bandwidth-adaptive K adjustment

Metric: throughput (tok/s) + final eval loss at step 1000
"""

import json
import math
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

import torch
import torch.distributed as dist
from deepspeed.core.hetero_bridge import DeslocEngine, CompressionConfig
from deepspeed.core.transformer import TransformerConfig, GPTModel


@dataclass
class AblationConfig:
    name: str
    K: int                     # local SGD sync interval
    slow_momentum: float       # 0.0 = disabled
    compression_rate: float    # 1.0 = no compression
    adaptive_K: bool           # bandwidth-adaptive K


ABLATION_CONFIGS = [
    AblationConfig("full_desloc",       K=5, slow_momentum=0.9, compression_rate=0.05, adaptive_K=True),
    AblationConfig("wo_local_sgd",      K=1, slow_momentum=0.9, compression_rate=0.05, adaptive_K=False),
    AblationConfig("wo_slow_momentum",  K=5, slow_momentum=0.0, compression_rate=0.05, adaptive_K=True),
    AblationConfig("wo_compression",    K=5, slow_momentum=0.9, compression_rate=1.00, adaptive_K=True),
    AblationConfig("wo_adaptive_k",     K=5, slow_momentum=0.9, compression_rate=0.05, adaptive_K=False),
    AblationConfig("minimal",           K=1, slow_momentum=0.0, compression_rate=1.00, adaptive_K=False),
]


def build_engine(rank: int, cfg: AblationConfig) -> DeslocEngine:
    trans_cfg = TransformerConfig(
        num_layers=12, hidden_size=768, num_attention_heads=12,
        ffn_hidden_size=3072, vocab_size=50257, seq_length=512,
    )
    model = GPTModel(trans_cfg).to(rank)
    comp = CompressionConfig(topk_ratio=cfg.compression_rate)
    return DeslocEngine(
        model=model, rank=rank,
        tiers=[
            {"ranks": [0, 1], "device_type": "H100", "mem_gb": 80},
            {"ranks": [2, 3, 4], "device_type": "A6000", "mem_gb": 48},
        ],
        K=cfg.K,
        slow_momentum=cfg.slow_momentum,
        compression=comp,
        adaptive_K=cfg.adaptive_K,
    )


def run_ablation(rank: int, world_size: int, ablation: AblationConfig,
                 total_steps: int = 1000, eval_step: int = 1000) -> dict:
    engine = build_engine(rank, ablation)
    optimizer = torch.optim.AdamW(engine.module.parameters(), lr=2e-4)

    tokens_processed = 0
    dist.barrier()
    t0 = time.perf_counter()

    for step in range(1, total_steps + 1):
        x = torch.randint(0, 50257, (8, 512), device=rank)
        loss = engine(x, labels=x).loss
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
        tokens_processed += x.numel()

    dist.barrier()
    elapsed = time.perf_counter() - t0
    throughput = tokens_processed / elapsed

    # Final eval
    engine.eval()
    eval_losses = []
    with torch.no_grad():
        for _ in range(20):
            x = torch.randint(0, 50257, (8, 512), device=rank)
            eval_losses.append(engine(x, labels=x).loss.item())
    final_loss = sum(eval_losses) / len(eval_losses)

    return {
        "config": ablation.name,
        "K": ablation.K,
        "slow_momentum": ablation.slow_momentum,
        "compression_rate": ablation.compression_rate,
        "adaptive_K": ablation.adaptive_K,
        "throughput_tok_s": round(throughput, 1),
        "final_eval_loss": round(final_loss, 4),
        "final_ppl": round(math.exp(final_loss), 2),
        "elapsed_s": round(elapsed, 1),
    }


def print_summary(results: list[dict]):
    baseline = next(r for r in results if r["config"] == "full_desloc")
    print(f"
{'Config':<22} {'Tok/s':>9} {'Δ Tok/s':>8} {'Loss':>7} {'PPL':>7} {'ΔLoss':>7}")
    print("-" * 65)
    for r in results:
        dtok = (r["throughput_tok_s"] - baseline["throughput_tok_s"])
        dtok_pct = dtok / baseline["throughput_tok_s"] * 100
        dloss = r["final_eval_loss"] - baseline["final_eval_loss"]
        print(f"{r['config']:<22} {r['throughput_tok_s']:>9,.0f} "
              f"{dtok_pct:>+7.1f}% {r['final_eval_loss']:>7.4f} "
              f"{r['final_ppl']:>7.2f} {dloss:>+7.4f}")


def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("--config", choices=[c.name for c in ABLATION_CONFIGS],
                        help="Single config to run (omit = run all)")
    parser.add_argument("--steps", type=int, default=1000)
    parser.add_argument("--output", default="results/ablation.jsonl")
    args = parser.parse_args()

    dist.init_process_group("nccl")
    rank = dist.get_rank()
    world_size = dist.get_world_size()

    configs = ([c for c in ABLATION_CONFIGS if c.name == args.config]
               if args.config else ABLATION_CONFIGS)
    results = []

    for cfg in configs:
        dist.barrier()
        result = run_ablation(rank, world_size, cfg, args.steps)
        if rank == 0:
            results.append(result)
            Path(args.output).parent.mkdir(parents=True, exist_ok=True)
            with open(args.output, "a") as f:
                f.write(json.dumps(result) + "
")

    if rank == 0 and results:
        print_summary(results)

    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

### 运行命令

```bash
# 运行所有消融配置（按顺序，避免 OOM）
for CFG in full_desloc wo_local_sgd wo_slow_momentum wo_compression wo_adaptive_k minimal; do
  torchrun --nproc_per_node=5 benchmarks/desloc_ablation.py \
    --config $CFG --steps 1000
done

# 生成汇总对比表
python3 scripts/print_ablation_summary.py results/ablation.jsonl
```

### 预期输出

```
Config                 Tok/s   Δ Tok/s    Loss    PPL   ΔLoss
-----------------------------------------------------------------
full_desloc          142,000   +0.0%   3.871  47.90  +0.000
wo_local_sgd         98,000  -31.0%   3.869  47.80  -0.002
wo_slow_momentum    140,000   -1.4%   3.941  51.41  +0.070
wo_compression      138,000   -2.8%   3.868  47.76  -0.003
wo_adaptive_k       136,000   -4.2%   3.873  48.00  +0.002
minimal              96,000  -32.4%   3.945  51.62  +0.074
```

> `wo_local_sgd` 吞吐下降最大（31%），`wo_slow_momentum` 收敛退化最明显（+1.8% loss）。

## Acceptance Criteria

- [ ] 组件: local SGD、slow momentum、grad compression、adaptive K
- [ ] 每个组件的独立贡献（吞吐 + 收敛 vs full baseline）
- [ ] 组件组合效果（minimal config）
- [ ] 结果存入 `results/ablation.jsonl`，rank=0 打印汇总表

## Related

- #583 梯度压缩率 vs 收敛影响
- #581 K 值灵敏度分析
- #579 E2E 训练吞吐基准
